morphir/runtime/src/org/finos/morphir/runtime/quick/Native.scala

```diff
@@ -85,144 +76,165 @@ object BasicsSDK {
-    (a: Result[Unit, Type.UType]) =>
-      Result.Primitive(!Result.unwrap(a).asInstanceOf[Boolean])
-  )
-  val plus: SDKValue[Unit, Type.UType] = SDKValue.SDKNativeFunction(
-    2,
-    (a: Result[Unit, Type.UType], b: Result[Unit, Type.UType]) =>
-      Result.Primitive(Result.unwrap(a).asInstanceOf[Long] + Result.unwrap(b).asInstanceOf[Long])
-  )
-  val subtract: SDKValue[Unit, Type.UType] = SDKValue.SDKNativeFunction(
-    2,
-    (a: Result[Unit, Type.UType], b: Result[Unit, Type.UType]) =>
-      Result.Primitive(Result.unwrap(a).asInstanceOf[Long] - Result.unwrap(b).asInstanceOf[Long])
+  private def handleSameNumerics(
+      a: Result[Unit, Type.UType],
+      b: Result[Unit, Type.UType]
+  )(f: (Any, Any, scala.Numeric[Any]) => Any): Result.Primitive[Unit, Type.UType, _] = {
+    val components = Result.unwrapNumericsWithHelper[Unit, Type.UType, Any](a, b)
+    Result.Primitive.makeOrFail[Unit, Type.UType, Any](f(components.a, components.b, components.helper))
+  }
+
+  private def handleBooleans(
+      a: Result[Unit, Type.UType],
+      b: Result[Unit, Type.UType]
+  )(f: (Boolean, Boolean) => Boolean) = {
+    val aBool = Result.unwrapBoolean(a)
+    val bBool = Result.unwrapBoolean(b)
+    Result.Primitive.Boolean[Unit, Type.UType](f(aBool, bBool))
+  }
+
+  val and: SDKValue[Unit, Type.UType] =
+    SDKValue.SDKNativeFunction.fun2 { (a: Result[Unit, Type.UType], b: Result[Unit, Type.UType]) =>
+      handleBooleans(a, b)(_ && _)
+    }
+  val or: SDKValue[Unit, Type.UType] =
+    SDKValue.SDKNativeFunction.fun2 { (a: Result[Unit, Type.UType], b: Result[Unit, Type.UType]) =>
+      handleBooleans(a, b)(_ || _)
+    }
+  val not: SDKValue[Unit, Type.UType] = SDKValue.SDKNativeFunction.fun1((a: Result[Unit, Type.UType]) =>
+    Result.Primitive.Boolean(!a.unwrapBoolean)
   )
-  val divide: SDKValue[Unit, Type.UType] = SDKValue.SDKNativeFunction(
-    2,
-    (a: Result[Unit, Type.UType], b: Result[Unit, Type.UType]) =>
-      Result.unwrap(a) match {
-        case l: Long =>
-          Result.Primitive(l / Result.unwrap(b).asInstanceOf[Long])
-        case l: Double =>
-          Result.Primitive(l / Result.unwrap(b).asInstanceOf[Double])
+  val plus: SDKValue[Unit, Type.UType] =
+    SDKValue.SDKNativeFunction.fun2 { (a: Result[Unit, Type.UType], b: Result[Unit, Type.UType]) =>
+      handleSameNumerics(a, b) { (aNum, bNum, helper) => helper.plus(aNum, bNum) }
+    }
+  val subtract: SDKValue[Unit, Type.UType] =
+    SDKValue.SDKNativeFunction.fun2 { (a: Result[Unit, Type.UType], b: Result[Unit, Type.UType]) =>
+      handleSameNumerics(a, b) { (aNum, bNum, helper) => helper.minus(aNum, bNum) }
+    }
+  val divide: SDKValue[Unit, Type.UType] =
+    SDKValue.SDKNativeFunction.fun2 { (a: Result[Unit, Type.UType], b: Result[Unit, Type.UType]) =>
+      handleSameNumerics(a, b) { (aNum, bNum, helper) =>
+        // scala.Numeric does not handle division, it's part of scala.Fractional or scala.Integral, need to get the right one
```

Comment:
`divide` as specified in morphir.SDK is only for floats - integerDivide is a completely different function